AssembleProgramming/AssembleScript
# Major Updates to Assemble Script

# PR Summary

This pull request contains several updates and improvements to the Assemble Script language. The updates include the addition of user-defined functions, increasing the max array size limit to 10,000,000, and fixing several bugs and broken code. The changes aim to enhance the functionality and reliability of the language.

## Changes Made

1. **User-Defined Functions:** Added support for user-defined functions in Assemble Script. Now users can define their functions to modularize their code and improve readability.

2. **Max Array Size Limit:** Increased the maximum size limit of arrays in Assemble Script to 10,000,000. This enhancement allows users to work with larger datasets and arrays.

3. **Bug Fixes:**
   - Fixed the bug causing issues with the `break` statement in loops.
   - Resolved the issue with the default return type of the if-else statement, which was undefined. Now the default return type is `null`.
   - Created a new runtime value type `boolean`, enabling boolean operations in the language.
   - Fixed the assignment to boolean type values.
   - Resolved the bug that prevented the declaration of variable-length arrays.
   - Addressed heap overflow caused by string multiplication.

## New User-Defined Functions Syntax
With the latest updates to Assemble Script, you can now define your own functions using the following syntax:
```js
def functionName(param1, param2, ...) {
    // Function body
    // You can use the provided parameters in your code
    // Function can have multiple statements
    // Return a value using the "return" keyword
}
```
## How to Test

To test the changes, please run the provided test code in the Assemble Script interpreter/compiler. The test code includes various scenarios that cover the newly added features and bug fixes. Ensure that all tests pass successfully.

## Additional Changes

Apart from the major updates mentioned above, this pull request includes minor code cleanups and optimizations to improve the overall code quality and performance.

## Types of Changes

What types of changes does your code introduce to this project? Please mark with
an `x` the applicable boxes.

- [x] Bugfix (non-breaking change that addresses an issue)


- [x] Breaking change (fix or feature that may cause existing functionality to
      break)


- [x] Refactoring (code optimization or restructuring)

- [ ] Other (please specify): __________

## Testing and Publishing

Please indicate the status of the following:

- [x] These changes have been thoroughly tested.

- [x] Code is properly formatted using command `deno fmt`
